lib/concurrent-ruby/concurrent/atomic/locals.rb

```diff
@@ -0,0 +1,184 @@
+require 'concurrent/constants'
+
+module Concurrent
+  # @!visibility private
+  # @!macro internal_implementation_note
+  #
+  # An abstract implementation of local storage, with sub-classes for
+  # per-thread and per-fiber locals.
+  #
+  # Each execution context (EC, thread or fiber) has a lazily initialized array
+  # of local variable values. Each time a new local variable is created, we
+  # allocate an "index" for it.
+  #
+  # For example, if the allocated index is 1, that means slot #1 in EVERY EC's
+  # locals array will be used for the value of that variable.
+  #
+  # The good thing about using a per-EC structure to hold values, rather than
+  # a global, is that no synchronization is needed when reading and writing
+  # those values (since the structure is only ever accessed by a single
+  # thread).
+  #
+  # Of course, when a local variable is GC'd, 1) we need to recover its index
+  # for use by other new local variables (otherwise the locals arrays could
+  # get bigger and bigger with time), and 2) we need to null out all the
+  # references held in the now-unused slots (both to avoid blocking GC of those
+  # objects, and also to prevent "stale" values from being passed on to a new
+  # local when the index is reused).
+  #
+  # Because we need to null out freed slots, we need to keep references to
+  # ALL the locals arrays, so we can null out the appropriate slots in all of
+  # them. This is why we need to use a finalizer to clean up the locals array
+  # when the EC goes out of scope.
+  class AbstractLocals
+    def initialize(name_prefix = :concurrent_locals)
+      @free = []
+      @lock = Mutex.new
+      @all_locals = {}
+      @next = 0
+
+      @name = :"#{name_prefix}_#{object_id}"
+    end
+
+    def synchronize
+      @lock.synchronize { yield }
+    end
+
+    if Concurrent.on_cruby?
+      def weak_synchronize
+        yield
+      end
+    else
+      alias_method :weak_synchronize, :synchronize
+    end
+
+    def next_index(target)
+      index = synchronize do
+        if @free.empty?
+          @next += 1
+        else
+          @free.pop
+        end
+      end
+
+      # When the target goes out of scope, we should free the associated index
+      # and all values stored into it.
+      ObjectSpace.define_finalizer(target, target_finalizer(index))
+
+      return index
+    end
+
+    def free_index(index)
+      weak_synchronize do
+        # This is O(N) number of ECs with locals... I'm concerned about the
+        # performance cost. But I don't see a better way to do it. Using
+        # `each_value` can blow up if @all_locals is modified during iteration.
+        # Maybe using a linked list for `@all_locals` would be better?
+        @all_locals.values.each do |locals|
+          locals[index] = nil
+        end
+
+        @free << index
```

Comment:
We should keep the original implementation's comments, notably here there was:
> `# free index has to be published after the arrays are cleared`

Reply:
Done.